### ISSUE TITLE
arangod process hangs when running too many unit tests

### ISSUE THREAD

**Issue by ilearnio** [arangodb/arangodb#1946]
I happened to have a test case (Mocha) in my code that, for some reason, didn't always passed, so in order to reproduce it I duplicated it around 500+ times, then ran tests again (via browser). But I didn't get any results, the request just hung http://take.ms/feiXi as well as the arangod process. I tried to quit it with Ctrl+C, `kill`, `kill -15`, but were only able to kill it forcibly with `kill -9`.

I was able to reproduce it by installing `aye-aye` service. There is a test case that should be duplicated in about 1000 lines in order for it to hang

**Comment by dothebart** [comment]
maybe increasing `--javascript.v8-contexts` can fix this? If you intend to have more foxx specific load, more paralell available V8 contexts may work better.

**Comment by ilearnio** [comment]
@dothebart I tried as you suggested, it didn't help. Anyways I don't really need to run that many tests at the moment, just wanted you to be aware

**Comment by dothebart** [comment]
ok, can you provide a short testcase howto reproduce this?

**Comment by ilearnio** [comment]
I was able to reproduce it by installing `aye-aye` service. There is a test case that should be duplicated in about 1000 lines in order for it to hang

**Comment by plassa-b** [comment]
I experienced the same bug. You can reproduce this easily :
```js
/*global describe, it */
'use strict';
const assert = require('assert');

describe('science', function () {
  for (var i = 0; i < 10000; ++i) {
    it('works', function () {
      assert.notEqual(true, false);
    });
  }
});
```

It really is important for me. I have like more than 300 tests that I need to run, and it fails at about 150. My tests are obviously more complex than that, but anyway, I'd appreciate some help.

**Comment by ilearnio** [comment]
I remember for me ~500 tests were enough for it to hang. Seems that the `it` function is quite heavy

**Comment by plassa-b** [comment]
I just edited my comment. In my case, tests don't pend, they just execute without a detailed response (and probably not all of them actually ran)

**Comment by dothebart** [comment]
Are you running this on a mac? it seems that V8 is doing especially bad here - we split some of our testcases since they would regulary hang on our mac node of the CI.

**Comment by plassa-b** [comment]
I'm running these on a Docker container on Linux, not mac

**Comment by ilearnio** [comment]
I am on Mac

**Comment by plassa-b** [comment]
Any news on this ? Testing is a really important point for my work, so it'd be great if it didn't fail to test

**Comment by pluma** [comment]
We're currently unable to reproduce this issue on Linux but can confirm it in OSX. We were not yet able to determine the exact source of the problem -- it may have to do with memory restrictions in V8 triggered by mocha's implementation of the test runner.

**Comment by dashdeep13** [comment]
I had created #3890 which is a duplicate of this. A key thing to note is that with every such call the arangodb process cpu utilization increases BY 50% - 56%,102%... and so on. Another thing is that I used the arangodb  starter to create a 3 node cluster on Mac and the tests still failed! (Meaning its definitely a Mac issue :) )Finally in my case I used a 240 test case file for a service which only contains it(as test bodies) and a single assert true statement inside.
